Add mTLS (x509) authentication method

kahowell opened the pull request:
Note: I added a script: `scripts/setup_devel_env` which generate certs and configuration if they don't exist. It may prove useful in trying to test this change. This script configures against samltest.id, and generates a unique SP entityId, so that turnpike developers don't overwrite each other's configs accidentally.

The approach I took was to go ahead and have nginx perform the authentication (verifying the client cert against a CA), and then pass the client cert subject and issuer to the turnpike service for authorization logic.

Also, note that I changed the auth order to try x509 first. The service falls back to SAML if the x509 header is not present. In this way, an endpoint can support either authentication method (SAML for humans, mTLS for services).

`x509` auth takes a Python expression which evaluates to `True` or `False`. It is passed a dictionary `x509` which contains two
attributes: `subject_dn` and `issuer_dn`, which can be used to further restrict which certs can be used (nginx already verifies the trust chain based on the configured CA file).

For example to restrict the endpoint to a certificate with the DN of `CN=test`, you could use `x509['subject_dn'] == 'CN=test'`.

Note that CRL and/or OCSP support should be configured in `NGINX_SSL_CONFIG` as needed.

To test, setup needed environment, start the services, and then execute

```
curl -k --cert services/nginx/certs/cert.pem --key services/nginx/certs/key.pem https://turnpike.example.com/api/turnpike/identity
```

and observe the resulting JSON document. Try a non-matching cert (any old cert will do), and verify that it receives 403.

kahowell commented:
FYI, I noticed that the error code you get with a failing cert from nginx (failed trust verification) is actually 400, not 403. With this PR 403 happens when trust succeeds, but the predicate fails.

Also, FYI, I've got a WIP refactor of this on top of #12 at https://github.com/kahowell/turnpike/tree/feature/mTLS_refactor. Will keep this PR open until #12 lands, and I'll just force push my refactor to this branch when it's all ready.

j00bar commented:
Okay, @kahowell - #12 is merged.

In the plugin interface that it brings, I have attributes for headers that plugins would want to pass along to the upstream API server - re-reading your PR, it would probably make sense to have something similar for headers that plugins want passed into the `/auth` service. Will work on that now.

j00bar commented:
The change in the above comment is now pending as #13.

kahowell commented:
@j00bar I rebased, and went ahead and squashed down (should be able to find the previous commits if desired, please reach out otherwise). #13 made it easy enough to isolate the nginx-terminated client cert bits to just docker-compose (aside from awkardness w/ variable interpolation - I had to use `${_dollarhack:-$}ssl_client_verify` instead of `$$ssl_client_verify` to get `podman-compose` to play nice with `$ssl_client_verify` and other vars).

kahowell commented:
Also, notes added to the commit from the rebases:

These all work:
- visit https://turnpike.example.com/api/turnpike/identity/ and login via SAML
- mimicking upstream TLS termination: `curl -k -H 'x-rh-certauth-cn: /CN=foo' -H 'x-rh-certauth-issuer: /CN=bar' https://turnpike.example.com/api/turnpike/identity/`
- having nginx do TLS termination: `curl -k --cert ./saml/certs/sp.crt --key ./saml/certs/sp.key https://turnpike.example.com/api/turnpike/identity/`

When the x-rh-certauth-cn and x-rh-certauth-issuer headers are passed, pass them down.
    
For development purposes, I'm still configuring nginx to perform client cert auth if it is passed.
    
Note that I noticed our upstream gateway's DN form starts with a '/', so I prepended this when client cert auth is used.
    
I assume x-rh-certauth-cn is a misnomer, as it looks to be the whole DN, starting with /CN=. It was likely named as CN since that is the only field in a RHSM-registered system's DN.
    
If this assumption is wrong, or if we introduce an additional header for the subject DN, should be a quick tweak to fix.

kahowell commented:
@j00bar, as discussed, moved to `x-rh-certauth-subject`, and made it configurable. Also made a couple of updates to docs.

j00bar commented on turnpike/plugins/x509.py:
```diff
@@ -0,0 +1,28 @@
+import logging
+from flask import request
+
+from ..config import HEADER_CERTAUTH_SUBJECT, HEADER_CERTAUTH_ISSUER
```
Doing the import of the config this way would prevent installers from overriding with their own config file. What you probably want to do is something like:

```
from flask import current_app

class X509AuthPlugin(TurnpikeAuthPlugin):
    # ... snip ...
    @property
    def headers_needed(self):
        return set([current_app.config.HEADER_CERTAUTH_SUBJECT, current_ap.config.HEADER_CERTAUTH_ISSUER])
```

Does that make sense?

j00bar reviewed (changes requested):
This is really awesome - I'm grateful for you writing it, including the helper script!

I'm going to pull this and test it locally, but otherwise it looks great.
